File: exercism/python/matrix/matrix.py
# https://exercism.io/my/solutions/2f6e880c131d4a35bdb83063bf5ff996


class Matrix(object):
    """2D array of integers."""

    def __init__(self, matrix_string):
        self._matrix = [[int(i) for i in s.split()]
                        for s in matrix_string.split('\n')]

    def row(self, index):
        """Returns the index-th row of the matrix."""

        if index > 0 and index <= len(self._matrix):
            return self._matrix[index-1]
        else:
            raise MatrixIndexError('Matrix index out of range.')

    def column(self, index):
        """Returns the index-th column of the matrix.

        It assumes all the rows have equal length,
        although input strings are not validated."""

        if index > 0 and index <= len(self._matrix[0]):
            return [self._matrix[j][index-1] for j in range(len(self._matrix))]
        else:
            raise MatrixIndexError('Matrix index out of range.')


class MatrixIndexError(Exception):
    pass

File: exercism/go/matrix/matrix.go
// https://exercism.io/my/solutions/0e464bb92c684c829edf73bc79a51058

package matrix

import (
	"fmt"
	"strconv"
	"strings"
)

// Matrix type represents a two-dimensional slice of integers
type Matrix [][]int

func (m Matrix) height() int {
	return len(m)
}

func (m Matrix) width() int {
	if m.height() > 0 {
		return len(m[0])
	}
	return -1
}

// New creates a Matrix from a string
func New(s string) (Matrix, error) {
	lines := strings.Split(s, "\n")
	m := make(Matrix, len(lines))
	for i, line := range lines {
		rowS := strings.Fields(line)
		length := len(rowS)
		if i > 0 && length != m.width() {
			return m, fmt.Errorf("invalid input %q (all rows must have equal size)", s)
		}
		rowI := make([]int, length)
		for j := 0; j < length; j++ {
			n, err := strconv.Atoi(rowS[j])
			if err != nil {
				return m, fmt.Errorf("invalid input %q (elements must be integers)", s)
			}
			rowI[j] = n
		}
		m[i] = rowI
	}
	return m, nil
}

// Set sets the value of an specific element of a Matrix (in-place)
func (m Matrix) Set(row, col, val int) bool {
	if row >= 0 && row < m.height() && col >= 0 && col < m.width() {
		m[row][col] = val
		return true
	}
	return false
}

// Rows returns a slice of a Matrix's rows
func (m Matrix) Rows() [][]int {
	copiedArray := make([][]int, m.height())
	for i, row := range m {
		copiedRow := make([]int, len(row))
		copy(copiedRow, row)
		copiedArray[i] = copiedRow
	}
	return copiedArray
}

// Cols returns a slice of a Matrix's cols
func (m Matrix) Cols() [][]int {
	transArray := make([][]int, m.width())
	for c := 0; c < m.width(); c++ {
		transRow := make([]int, m.height())
		for r := 0; r < m.height(); r++ {
			transRow[r] = m[r][c]
		}
		transArray[c] = transRow
	}
	return transArray
}
